pkg/channel/lookup.go

```diff
@@ -133,26 +149,46 @@ type Lookup interface {
 	WithContext(ctx context.Context) (*ModuleTemplate, error)
 }
 
+func NewIndexedTemplateLookup(client client.Reader, module operatorv1beta1.Module,
+	defaultChannel string,
+) *TemplateLookup {
+	return &TemplateLookup{
+		reader:         client,
+		readerIndexed:  true,
+		module:         module,
+		defaultChannel: defaultChannel,
+	}
+}
+
 func NewTemplateLookup(client client.Reader, module operatorv1beta1.Module,
 	defaultChannel string,
 ) *TemplateLookup {
 	return &TemplateLookup{
 		reader:         client,
+		readerIndexed:  false,
 		module:         module,
 		defaultChannel: defaultChannel,
 	}
 }
 
 type TemplateLookup struct {
 	reader         client.Reader
+	readerIndexed  bool
 	module         operatorv1beta1.Module
 	defaultChannel string
 }
 
 func (c *TemplateLookup) WithContext(ctx context.Context) (*ModuleTemplate, error) {
 	desiredChannel := c.getDesiredChannel()
 
-	template, err := c.getTemplate(ctx, desiredChannel)
+	var template *operatorv1beta1.ModuleTemplate
+	var err error
+	if c.readerIndexed {
+		template, err = c.getTemplate(ctx, desiredChannel)
```

Comment:
If I get your point currently, the concern here to treat differently is because this index field can't be used in remote module template? `err := c.reader.List(ctx, templateList, option, index.TemplateChannelField.WithValue(desiredChannel))`

Let's have a sync regarding this problem, and if it's true, then we should concern about using a unified way to get template, maybe you solution is more generic, introducing more different solutions means more potential divergence and more maintenance efforts, and we should definitely avoid it.